qiita_db/analysis.py

```diff
@@ -808,81 +808,113 @@ def build_files(self,
 
-            # this assumes that there is only one reference/pipeline for each
-            # data_type issue #164
-            new_tables = {dt: None for dt in self.data_types}
-            for aid, samps in viewitems(samples):
-                artifact = qdb.artifact.Artifact(aid)
-                # this is not checking the reference used for picking
-                # issue #164
-                biom_table_fp = None
-                for _, fp, fp_type in artifact.filepaths:
-                    if fp_type == 'biom':
-                        biom_table_fp = fp
-                        break
-                if not biom_table_fp:
-                    raise RuntimeError(
-                        "Artifact %s do not have a biom table associated"
-                        % aid)
-                biom_table = load_table(biom_table_fp)
-                # filtering samples to keep those selected by the user
-                biom_table_samples = set(biom_table.ids())
-                selected_samples = biom_table_samples.intersection(samps)
-                biom_table.filter(selected_samples, axis='sample',
-                                  inplace=True)
-
-                if rename_dup_samples:
-                    ids_map = {_id: "%d.%s" % (aid, _id)
-                               for _id in biom_table.ids()}
-                    biom_table.update_ids(ids_map, 'sample', True, True)
+            _, base_fp = qdb.util.get_mountpoint(self._table)[0]
+            for label, tables in viewitems(grouped_samples):
+                data_type, reference_id, command_id = label.split('.')
+                new_table = None
+                artifact_ids = []
+                for aid, samples in tables:
+                    artifact = qdb.artifact.Artifact(aid)
+                    artifact_ids.append(str(aid))
+
+                    # the next loop is assuming that an artifact can have only
+                    # one biom, which is a safe assumption until we generate
+                    # artifacts from multiple bioms and even then we might
+                    # only have one biom
+                    biom_table_fp = None
+                    for _, fp, fp_type in artifact.filepaths:
+                        if fp_type == 'biom':
+                            biom_table_fp = fp
+                            break
+                    if not biom_table_fp:
+                        raise RuntimeError(
+                            "Artifact %s do not have a biom table associated"
+                            % aid)
+
+                    # loading the found biom table
+                    biom_table = load_table(biom_table_fp)
+                    # filtering samples to keep those selected by the user
+                    biom_table_samples = set(biom_table.ids())
+                    selected_samples = biom_table_samples.intersection(samples)
+                    biom_table.filter(selected_samples, axis='sample',
```

Comment:
Although very unlikely, this could return an empty table, which would still be a valid object, but would otherwise useless for everything processing related right? Do we want to do any check for that, maybe once we have the final "merged" table?